libsolidity/analysis/DeclarationTypeChecker.cpp

```diff
@@ -297,12 +297,18 @@ void DeclarationTypeChecker::endVisit(ArrayTypeName const& _typeName)
 		else if (optional<ConstantEvaluator::TypedRational> value = ConstantEvaluator::evaluate(m_errorReporter, *length))
 			lengthValue = value->value;
 
-		if (!lengthValue || lengthValue > TypeProvider::uint256()->max())
```

Comment:
Apparently the comparison in this old check is missing an asterisk, when I just copy-pasted this, an edge case failed (`arr[2**256-1]`) when it shouldn't. I added this as a regression test too

Reply:
> somehow 2**28 as a rational is considered bigger than uint256()->max()

Thanks for reporting this. Could you please open this as a separate issue, perhaps it has implications elsewhere in the code?

Reply:
> This is fixed once I use lengthValue->numerator(). It's safe because there is already a check for lengthValue->denominator() == 1

Where is this check exactly?

Reply:
> somehow `2**28` as a rational is considered bigger than `uint256()->max()`

Sounds like it might be doing a signed comparison - with `uint256()->max()` getting converted to `int256`, giving -1.

Reply:
> > somehow 2**28 as a rational is considered bigger than uint256()->max()
> 
> Thanks for reporting this. Could you please open this as a separate issue, perhaps it has implications elsewhere in the code?

Sorry it's a false alarm. It's an error from `ArrayType::validForLocation` which makes sure memory data isn't larger than `2^32`. It doesn't affect the type checker here. Let me know if you want me to use the old way for comparison here.

Reply:
Yes, I'd revert to the old one if it works because it's more straightforward (does not require a sanity check or any special explanations).